BaseStation/MainWindow/ConsoleView.xaml.cs

```diff
@@ -34,32 +33,70 @@ private class ConsoleWriter : TextWriter {
 
  			public override Encoding Encoding => Encoding.UTF8;
 
-			public override void Write(char value) {
+            public override void Write(char value)
+            {
                 oldOut.Write(value);
-				view.Dispatcher.BeginInvoke(new Action(() => {
-					view.Text += value;
-					UpdateView();
-				}));
-			}
+                view.Dispatcher.BeginInvoke(new Action(() =>
+                {
+                    view.Text += value;
+                    UpdateView();
+                }));
+            }
 
-			public override void Write(string value)
+            public override void Write(char[] buffer, int index, int count)
             {
-                oldOut.Write(value);
-                view.Dispatcher.BeginInvoke(new Action(() => {
-					view.Text += value;
-					UpdateView();
-				}));
-			}
+                oldOut.Write(buffer, index, count);
+                view.Dispatcher.BeginInvoke(new Action(() =>
+                {
+                    view.Text += new string(buffer, index, count);
+                    UpdateView();
+                }));
+            }
 
-			//removes extra text (prevent memory leak)
-			//and scrolls view to bottom (like a console)
-			private void UpdateView() {
+            //removes extra text (prevent memory leak)
+            //and scrolls view to bottom (like a console)
+            private void UpdateView() {
 				int len = view.Text.Length;
-				int removeSize = Math.Max(0, Math.Min(len - BUFFER_SIZE, BUFFER_SIZE));
-				view.Text =
-					view.Text.Substring(removeSize, len - removeSize);
+                double scrollHeight = scroll.ViewportHeight;
+                double viewHeight = view.ActualHeight;
+                int totalLines = CountLines(view.Text);
+                int lineHeight = (int)viewHeight/totalLines;
+                int possibleVisibleLines = (int)(scrollHeight/lineHeight);
+                if (viewHeight > scrollHeight * 2) //if stored text is partially offscreen, the 2 is there to buffer the text so there isn't so much garbage produced
+                {
+                    view.Text = GetFinalLines(view.Text, totalLines, Math.Min(totalLines, possibleVisibleLines));
+                }
 				scroll.ScrollToEnd();
 			}
-		}
+
+            private static string GetFinalLines(string text, int totalLines, int lineCount)
+            {
+                int ind = GetIndexAfter(text, totalLines - lineCount);
+                return text.Substring(ind);
+            }
+
+            private static int GetIndexAfter(string str, int count)
+            {
+                if (str == string.Empty)
+                    return -1;
```

Comment:
Braces would make these sections more readable, you can compress it all to a single line if you don't want to take up more vertical space.